### ISSUE TITLE
[bug][1.2] unstructure hook not applied when use 'from __future__ import annotations'

### ISSUE THREAD

**Issue by Bryant-Yang** [Tinche/cattrs#119]
version: cattrs==1.2;
python: 3.8

Hi, the test code as below, if uncomment first line, test failed. And I really need 'from __future__ import annotations' for my mypy checking in somewhere.

Another issue is mypy checking error on the same code below, such as:
error: Cannot infer type argument 1 of "register_unstructure_hook" of "Converter"

```
# from __future__ import annotations  # uncomment this line test will fail

from datetime import datetime
import cattr
from attr import attrs, ib

@attrs
class TestData:
    dt: datetime = ib()

converter = cattr.Converter()
converter.register_unstructure_hook(datetime,
                                    lambda val: val.strftime('%Y-%m-%d %H:%M:%S')
                                    if val is not None else None)

data = TestData(datetime.now())

result = converter.unstructure(data)

print(type(result.get('dt')))
assert type(result.get('dt')) == str
```

**Comment by Tinche** [comment]
Hi,

you can try using `cattr.GenConverter` instead and it should work. `GenConverter` will become the default converter in the future.

**Comment by Bryant-Yang** [comment]
GenConverter works fine for the simple test code above. But custom hook is still not triggered in my project (more complex context). I'm doing more inspection.

**Comment by Bryant-Yang** [comment]
well, it caused by the line 'from __future__ import annotations' too, tested on some data model in my real project, strange. The essential logic should be same as the test code.

**Comment by Bryant-Yang** [comment]
It caused by previous workaround on custom hidden fields: https://github.com/Tinche/cattrs/issues/99, I'll figure out another workaround.

**Comment by Tinche** [comment]
Interesting, can you put together a small example? I don't understand it completely but I feel like it should work.

**Comment by Bryant-Yang** [comment]
ok, use this test:

```
from __future__ import annotations # comment this line, test will pass

from datetime import datetime
import cattr
from attr import attrs


@attrs(auto_attribs=True, kw_only=True)
class BaseData:
    dt: Optional[datetime] = None


@attrs(auto_attribs=True, kw_only=True)
class TestData(BaseData):
    foo: str = 'foo'
    bar: str = 'bar'


converter = cattr.GenConverter()
converter.register_unstructure_hook(datetime,
                                    lambda val: val.strftime('%Y-%m-%d %H:%M:%S')
                                    if val is not None else None)

hidden_field_names = ['bar']

converter.register_unstructure_hook(
    TestData,
    lambda instance: {
        k: v
        for k, v in converter.unstructure_attrs_asdict(instance).items()
        if k not in hidden_field_names
    }
)



data = TestData(dt=datetime.now())


result = converter.unstructure(data)

print(result)
print(type(result.get('dt')))
assert type(result.get('dt')) == str


```

**Comment by Tinche** [comment]
Hm interesting.

The easiest way to solve this is by adding `attr.resolve_types(TestData)` before the `register_unstructure_hook` call.

A more general solution would be for cattrs to call this on your class when registering un/structuring hooks. I will look into how to handle this.

In the next release this use case should be easier to achieve. We can leave this open until that is released.

**Comment by joeldodge79** [comment]
I think I've hit the same problem. Switching to `GenConverter` and calling `attr.resolve_types` got me close. However, my unstructure hook is still being ignored (applying `register_unstructure_hook` afterwards as @Tinche suggests).

@Bryant-Yang did this work around unblock you? If so then I'll produce a simplified version of my code (perhaps in another issue) to see if anyone can see what I need to do differently. For now I gotta pin to 1.1.2

**Comment by Tinche** [comment]
@joeldodge79 I'd be curious for a small repro example that's not fixed by attr.resolve_types.

**Comment by Bryant-Yang** [comment]
@joeldodge79 my test passed by using attr.resolve_types.

**Comment by joeldodge79** [comment]
thanks for the info @Bryant-Yang 

@Tinche I'm deep in another project right now and it always takes me a minute to reload our project's cattrs usage (models/methods autogenerated from an OAS REST spec on top of a hand-written transport/auth/[de]serialization library).

Note to self: I bet the problem is [here](https://github.com/looker-open-source/sdk-codegen/blob/00bf83d0f4ad99c527c4fc2281c0616af000d49f/python/tests/rtl/test_serialize.py#L121): we hacked `__annotations__` to set types to `ForwardRef("foo")` values and I wonder how that plays with `attr.resolve_types`

We're gonna need to support 3.6 for a while and pinning at cattrs==1.1.2 is working. It seems like the `GenConverter` came in the [same release 3.6 support was dropped](https://github.com/Tinche/cattrs/blob/master/HISTORY.rst#110-2020-10-29) and I'd like to minimize the amount of `compat` code we need to maintain.

That being said I do want to make sure we can update when we're able to drop 3.6 support so it's on my list to dig into why our usage isn't working with the `attr.resolve_types` workaround in case it uncovers another case for you to handle going forward. But for now feel free to close this issue and I can open a new one (referencing this conversation) when I get a few minutes.

**Comment by Tinche** [comment]
Ah, we don't really support forward references (the idea is for you to use normal annotations and the future import to get that functionality), so that might be it.

**Comment by joeldodge79** [comment]
I'll try switching to the annotations future import to see if that solves our problem. But from what I understand it isn't available in 3.6 so I guess we're staying on 1.1.2 till we get rid of 3.6.

**Comment by Tinche** [comment]
`cattrs` now (since 1.3) calls `resolve_types` when registering hooks, so that should help.

**Comment by sobolevn** [comment]
I run into this issue:

```python
from __future__ import annotations

import cattr
import attr
from typing import List


@attr.dataclass(slots=True, frozen=True)
class Some(object):
    a: List[str]


cattr.structure({'a': ['1', '2', '3']}, Some)
```

Traceback:

```
  File "/Users/sobolev/Documents/github/typed-linter/.venv/lib/python3.7/site-packages/cattr/converters.py", line 205, in structure
    return self._structure_func.dispatch(cl)(obj, cl)
  File "/Users/sobolev/Documents/github/typed-linter/.venv/lib/python3.7/site-packages/cattr/converters.py", line 342, in structure_attrs_fromdict
    dispatch(type_)(val, type_) if type_ is not None else val
  File "/Users/sobolev/Documents/github/typed-linter/.venv/lib/python3.7/site-packages/cattr/converters.py", line 286, in _structure_default
    raise ValueError(msg)
ValueError: Unsupported type: List[str]. Register a structure hook for it.
```

Versions:
- attrs: 20.3.0
- cattrs: 1.4.0
- python: 3.7.9

**Comment by sobolevn** [comment]
Yes, this works! Thanks!

**Comment by natemcmaster** [comment]
I happened to just hit this myself. Thanks @Tinche for providing an answer!

Side note: with `from __future__ import annotations` [becoming the default behavior in Python 3.10](https://docs.python.org/3.10/whatsnew/3.10.html#pep-563-postponed-evaluation-of-annotations-becomes-default), I assume this will become a more common error later this year. Is there already an issue for tracking making GenConverter the default behavior?

**Comment by Tinche** [comment]
I'm kind of letting folks test it out in the wild for the time being. I would say the release after the next one will be 2.0 and have the GenConverter be the default converter.

**Comment by alexchandel** [comment]
@Tinche Could you please put a message to detect and warn against this in `Converter`?  The present "silent failure" is difficult to debug.

**Comment by Tinche** [comment]
@alexchandel The new GenConverter is the default nowadays, so it should work out of the box? Also it looks like Python will be moving away from this approach after all.
